dubbo-configcenter/dubbo-configcenter-nacos/src/main/java/org/apache/dubbo/configcenter/support/nacos/NacosDynamicConfiguration.java

```diff
@@ -0,0 +1,255 @@
+     *
+     * @param key   property key the native listener will listen on
+     * @param group to distinguish different set of properties
+     * @return
+     */
+    private NacosConfigListener createTargetListener(String key, String group) {
+        NacosConfigListener configListener = new NacosConfigListener();
+        configListener.fillContext(key, group);
+        return configListener;
+    }
+
+    @Override
+    public void addListener(String key, String group, ConfigurationListener listener) {
+        NacosConfigListener nacosConfigListener = watchListenerMap.computeIfAbsent(generateKey(key, group), k -> createTargetListener(key, group));
+        String keyInNacos = rootPath + PROPERTIES_CHAR_SEPERATOR + key;
+        nacosConfigListener.addListener(listener);
+        try {
+            configService.addListener(keyInNacos, null, nacosConfigListener);
+        } catch (NacosException e) {
+            logger.error(e.getMessage());
+        }
+    }
+
+    private String generateKey(String key, String group) {
+        if (StringUtils.isNotEmpty(group)) {
+            key = key + GROUP_CHAR_SEPERATOR + group;
+        }
+        return key;
+    }
+
+    @Override
+    public void removeListener(String key, String group, ConfigurationListener listener) {
+        NacosConfigListener eventListener = watchListenerMap.get(generateKey(key, group));
+        if (eventListener != null) {
+            eventListener.removeListener(listener);
+        }
+    }
+
+    @Override
+    public String getConfig(String key, String group, long timeout) throws IllegalStateException {
+        key = generateKey(key, group);
+        return (String) getInternalProperty(rootPath + PROPERTIES_CHAR_SEPERATOR + key);
+    }
+
+    @Override
+    public Object getInternalProperty(String key) {
+        try {
+            String[] keyAndGroup = getKeyAndGroup(key);
+            return configService.getConfig(keyAndGroup[0], keyAndGroup[1], 5000L);
+        } catch (NacosException e) {
+            logger.error(e.getMessage());
+        }
+        return null;
+    }
+
+    public class NacosConfigListener extends AbstractSharedListener {
+
+        private Set<ConfigurationListener> listeners = new CopyOnWriteArraySet<>();
+        /**
+         * cache data to store old value
+         */
+        private Map<String, String> cacheData = new ConcurrentHashMap<>();
+
+        @Override
+        public Executor getExecutor() {
+            return null;
+        }
+
+        /**
+         * receive
+         *
+         * @param dataId     data ID
+         * @param group      group
+         * @param configInfo content
+         */
+        @Override
+        public void innerReceive(String dataId, String group, String configInfo) {
+            String oldValue = cacheData.get(dataId);
+            ConfigChangeEvent event = new ConfigChangeEvent(dataId, configInfo, getChangeType(configInfo, oldValue));
+            cacheData.put(dataId, configInfo);
```

Comment:
if `configInfo` is null, which means a config has been deleted. This line will throw NPE:

```java
        java.lang.NullPointerException
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
	at org.apache.dubbo.configcenter.support.nacos.NacosDynamicConfiguration$NacosConfigListener.innerReceive(NacosDynamicConfiguration.java:229)
	at com.alibaba.nacos.api.config.listener.AbstractSharedListener.receiveConfigInfo(AbstractSharedListener.java:37)
	at com.alibaba.nacos.client.config.impl.CacheData$1.run(CacheData.java:188)
	at com.alibaba.nacos.client.config.impl.CacheData.safeNotifyListener(CacheData.java:209)
	at com.alibaba.nacos.client.config.impl.CacheData.checkListenerMd5(CacheData.java:160)
	at com.alibaba.nacos.client.config.impl.ClientWorker$LongPollingRunnable.run(ClientWorker.java:505)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

In order to avoid that, it is necessary to remove the dataID from cacheDate.